tls/s2n_cipher_suites.c

```diff
@@ -934,6 +934,16 @@ static struct s2n_cert_chain_and_key *s2n_get_compatible_cert_chain_and_key(stru
     return NULL;
 }
 
+static struct s2n_cert_chain_and_key *s2n_conn_get_compatible_cert_chain_and_key(struct s2n_connection *conn, struct s2n_cipher_suite *cipher_suite)
+{
+    if (conn->handshake_params.num_sni_matching_certs > 0) {
+        return s2n_get_compatible_cert_chain_and_key(conn->handshake_params.sni_matching_certs, conn->handshake_params.num_sni_matching_certs, cipher_suite);
+    } else {
+        /* We don't have any name matches. Use the first certificate that works with the key type. */
+        return s2n_get_compatible_cert_chain_and_key(conn->config->cert_and_key_pairs, conn->config->num_certificates, cipher_suite);
```

Comment:
Checking my understanding: By this logic and the way `s2n_set_cipher_and_cert_as_server` works, if we have:
- cert A: ECC key, SAN for www.foo.com
- cert B: RSA key, SAN for www.bar.com
and the client requests server name www.foo.com but presents both ECDSA and RSA cipher suites, we will choose cert B (www.bar.com) if the server is configured to prefer RSA cipher suites over ECDSA cipher suites? I would have thought it would be better to return cert A in that case since it can satisfy all of the requirements (server name, client cipher prefs, server cipher prefs).

Reply:
I think this is the current behavior since `cert A` will be the only entity in `sni_matching_certs`. I'll add a unit test to confirm this. This test: https://github.com/awslabs/s2n/blob/32096ebb810eb0137241a490bcdbfc4ca21e9fd3/tests/integration/s2n_test_constants.py#L182-L184 *almost* matches the case you describe but the ECDSA cipher suite is higher on the server's preference so we don't get to verify that the server skips the "non sni matched" certificate.

Reply:
Added a test specifically for this case in 
e675624